h2o-core/src/main/java/hex/grid/filter/DefaultPermutationFilter.java

```diff
@@ -0,0 +1,38 @@
+package hex.grid.filter;
+
+import hex.Model;
+
+import java.util.List;
+
+public class DefaultPermutationFilter<MP extends Model.Parameters> implements PermutationFilter<MP> {
+
+  private List<PermutationFilterFunction<MP>> _filterFunctions;
+
+  public DefaultPermutationFilter(List<PermutationFilterFunction<MP>> filterFunctions) {
+    _filterFunctions = filterFunctions;
+  }
+  
+  @Override
+  public boolean permutationIsSkipped(MP permutation) {
+    boolean skipPermutation = false;
+
+    for (PermutationFilterFunction<MP> fun : _filterFunctions) {
+      if (!fun.apply(permutation)) {
+        skipPermutation = true;
+        break;
+      }
+    }
+
+    if (!skipPermutation) {
+      _filterFunctions.stream()
+              .filter(ff -> ff.test(permutation))
+              .forEach(Activatable::activate);
+    }
```

Comment:
thanks for allowing that all `PermutationFilterFunctions` behave the same way at this level.

Reply:
Experimenting with a `positive` version of this logic, it's much easier to reason about.
```
@Override
  public boolean test(MP permutation) {
    boolean match = true;

    for (FilterFunction<MP> fun : _filterFunctions) {
      if (!fun.apply(permutation)) {
        match = false;
        break;
      }
    }
    _filterFunctions.stream()
              .forEach(ff.accept(match, permutation));
    return match
}
```
Also, by calling both `test` and `activate` above, you're somehow delegating the activation logic to the caller, when it should be to the filter itself to decide. So I'd consider calling a single method like `accept(globalMatch, permutation)` which would just update its internal state based on the given parameters.

It would also clarify the logic in your `KeepFirstMatchFilter`:
```
  @Override
  public Boolean apply(MP permutation) {
    if(_maxNumberOfMatchesToApply == 0) {
      return  !test(permutation);
    } else {
      return true;
    }
  }

  @Override
  public boolean test(MP permutation) {
    return _baseMatchFunction.apply(permutation);
  }

  @Override
  public void accept(boolean globalMatch, MP permutation) {
    if (globalMatch && test(permutation) && _maxNumberOfMatchesToApply > 0) _maxNumberOfMatchesToApply--;
  }
```
Now it is clear that globalMatch is also a condition for the "activation" of this filter.
Also, "accept" allows you to get rid of the idea of "activation", as if the filter was not really applied before... it was, but it was just behaving differently.
Alternatives to `accept`, I suggest `track`, `record`...

Reply:
Again, instead of `activate` + `test` functions, this `accept` (or equivalent) could have a default no-op implementation in the functional interface

Reply:
@seb-h2o
> Also, "accept" allows you to get rid of the idea of "activation", as if the filter was not really applied before... it was, but it was just behaving differently.

 Better to have explicit separation of states. "it was just behaving differently" - sounds like a bad idea.

Reply:
@seb-h2o also `test` method is coming from Predicate interface that you suggested before. Will not be able to use it.

Reply:
> "it was just behaving differently" - sounds like a bad idea.

well, that's what your filter does, it behaves differently before and after activation.
I'm just suggesting to move all this decision logic into the filter itself, instead of having half of it inside, half outside.

Reply:
@seb-h2o >well, that's what your filter does, it behaves differently before and after activation.

but I had clear explicit transition from one state to another.

Reply:
@seb-h2o We need `test` method that does not depend on anything except permutation. Your method `accept` returns `void` so I don't see where we should take basic match logic from. See following method. `apply` is for high level test when we have stateful logic like from KeepFirstMatch filter. This apply method should for stateless filters call base `test` logic - but we will not have one with you suggestion.

```
default boolean apply(MP permutation) {
    return test(permutation);
  }
```
 I understand your intention but it will not work like you are suggesting. The only thing I can probably improve here is to move `.filter(ff -> ff.test(permutation))` logic inside `activate` method.